// nestjs-mailer/dist/mailer.core-module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var MailerCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const mailer_utils_1 = require("./mailer.utils");
let MailerCoreModule = MailerCoreModule_1 = class MailerCoreModule {
    static forRoot(options, connection) {
        const mailerOptionsProvider = {
            provide: mailer_utils_1.getMailerOptionsToken(connection),
            useValue: options,
        };
        const mailerConnectionProvider = {
            provide: mailer_utils_1.getMailerConnectionToken(connection),
            useValue: mailer_utils_1.createMailerConnection(options),
        };
        return {
            module: MailerCoreModule_1,
            providers: [
                mailerOptionsProvider,
                mailerConnectionProvider,
            ],
            exports: [
                mailerOptionsProvider,
                mailerConnectionProvider,
            ],
        };
    }
    static forRootAsync(options, connection) {
        const mailerConnectionProvider = {
            provide: mailer_utils_1.getMailerConnectionToken(connection),
            useFactory(options) {
                return mailer_utils_1.createMailerConnection(options);
            },
            inject: [mailer_utils_1.getMailerOptionsToken(connection)],
        };
        return {
            module: MailerCoreModule_1,
            imports: options.imports,
            providers: [...this.createAsyncProviders(options, connection), mailerConnectionProvider],
            exports: [mailerConnectionProvider],
        };
    }
    static createAsyncProviders(options, connection) {
        if (!(options.useExisting || options.useFactory || options.useClass)) {
            throw new Error('Invalid configuration. Must provide useFactory, useClass or useExisting');
        }
        if (options.useExisting || options.useFactory) {
            return [
                this.createAsyncOptionsProvider(options, connection)
            ];
        }
        return [
            this.createAsyncOptionsProvider(options, connection),
            { provide: options.useClass, useClass: options.useClass },
        ];
    }
    static createAsyncOptionsProvider(options, connection) {
        if (!(options.useExisting || options.useFactory || options.useClass)) {
            throw new Error('Invalid configuration. Must provide useFactory, useClass or useExisting');
        }
        if (options.useFactory) {
            return {
                provide: mailer_utils_1.getMailerOptionsToken(connection),
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        return {
            provide: mailer_utils_1.getMailerOptionsToken(connection),
            async useFactory(optionsFactory) {
                return await optionsFactory.createMailerModuleOptions();
            },
            inject: [options.useClass || options.useExisting],
        };
    }
};
MailerCoreModule = MailerCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({})
], MailerCoreModule);
exports.MailerCoreModule = MailerCoreModule;
